### DISCUSSION TITLE
How do I type the `persist` middleware `StorePersist` type?

### DISCUSSION THREAD

**Discussion by JDaoMothership** [pmndrs/zustand#2871] [Q&A] [answered]
Originally I had:

```typescript
import type { StoreApi } from 'zustand'
...
export const createMyStore = (customParams: MyParams): StoreApi<MyStore> => {
  return createStore<MyStore>()(
     ...
  )
}
```

And now I am trying to add the `persist` middleware:

```typescript
import type { StoreApi } from 'zustand'
...
export const createMyStore = (customParams: MyParams): StoreApi<MyStore> => {
  return createStore<MyStore>()(
     persist(
         ...
     )
  )
}
```

But `StoreApi` does not have the `persist` object in the return value. 

And I see the store type is now defined as `Write<StoreApi<MyStore>, StorePersist<MyStore, MyStore>>`

but `Write` and `StorePersist` are not exported as types from the `zustand/middleware` import, so I am not sure how best to type the return function here! Am I missing the best practice for this? 

I am on zustand version 4.5.2.

**Comment by dbritto-dev** [comment]
@JDaoMothership https://zustand.docs.pmnd.rs/guides/typescript#create-without-curried-workaround

**Reply by JDaoMothership** [reply]
This is not exactly what I need - basically in that example, I need to define a type for `useBearStore`. How would I do that?

**Reply by dbritto-dev** [reply]
@JDaoMothership in a nutshell just remove the return type

```ts
export const createMyStore = (customParams: MyParams) => {
  return createStore<MyStore>()(
     persist(
         ...
     )
  )
}
```

**Reply by JDaoMothership** [reply]
That does work - but I need to pass around the return from `createMyStore` into some components, and I'd like to type that correctly with the persist methods, and not just `StoreApi<MyStore>`

**Reply by dbritto-dev** [reply]
In that case you could use `ReturnType<typeof createMyStore>`

**Reply by JDaoMothership** [reply]
Is there no other way to do that? Are `Write` and `StorePersist` exposed in future versions of zustand?

**Reply by dbritto-dev** [reply]
@JDaoMothership those wouldn't be exposed in the future

**Reply by JDaoMothership** [reply]
Is there a reason for that? I feel like exposing those types here makes sense for this use case, right? Or am I approaching this incorrectly somehow?

**Reply by dbritto-dev** [reply]
@JDaoMothership those are more like for library authors that's why we didn't expose them

**Reply by JDaoMothership** [reply]
I understand, but I guess I'm wondering if this use case is valid, wouldn't it make sense to expose them? Or if it's not valid, I'm wondering what a better way to approach this is! :)

**Reply by dbritto-dev** [reply]
@JDaoMothership the best approach is to infer types as much as possible, explicit types could be tricky and not easy to maintain

**Comment by JDaoMothership** [comment, marked as answer]
Answer from @dbritto-dev here: https://github.com/pmndrs/zustand/discussions/2871#discussioncomment-11396748

> In that case you could use `ReturnType<typeof createMyStore>`
